File: velocity/src/main/java/bluesea/aquautils/PluginCommand.java
package bluesea.aquautils;

import bluesea.aquautils.common.Controller;
import com.velocitypowered.api.command.SimpleCommand;
import com.velocitypowered.api.proxy.Player;
import java.util.List;
import java.util.stream.Collectors;
import net.kyori.adventure.audience.Audience;
import net.kyori.adventure.text.Component;

public class PluginCommand implements SimpleCommand {
    @Override
    public void execute(Invocation invocation) {
        if (invocation.source() instanceof Player) {
            Audience players = Audience.audience(AquaUtilsVelocity.getServer().getAllPlayers());
            Audience player = invocation.source();

            if (invocation.alias().equals("vkick")) {
                if (invocation.arguments().length > 0) {
                    for (Player kickPlayer : AquaUtilsVelocity.getServer().getAllPlayers()) {
                        if (kickPlayer.getUsername().equals(invocation.arguments()[0])) {
                            kickPlayer.disconnect(
                                Component.translatable("multiplayer.disconnect.kicked"));
                            break;
                        }
                    }
                }
                return;
            }
            Controller.onCommand(players, player, invocation.alias(), invocation.arguments());
        }
    }

    @Override
    public List<String> suggest(Invocation invocation) {
        if (invocation.alias().equals("vkick") && invocation.arguments().length <= 1) {
            return AquaUtilsVelocity.getServer().getAllPlayers().stream().map(Player::getUsername).collect(Collectors.toList());
        }
        return SimpleCommand.super.suggest(invocation);
    }

    @Override
    public boolean hasPermission(Invocation invocation) {
        return invocation.source().hasPermission("aquautils." + invocation.alias());
    }
}

File: velocity/src/main/kotlin/bluesea/aquautils/PluginCommand.kt
package bluesea.aquautils

import bluesea.aquautils.common.Constants.ID
import bluesea.aquautils.parser.VelocityPlayers
import bluesea.aquautils.parser.VelocityPlayersParser
import com.velocitypowered.api.proxy.Player
import kotlin.jvm.optionals.getOrNull
import net.kyori.adventure.text.Component
import org.incendo.cloud.context.CommandContext
import org.incendo.cloud.parser.standard.StringParser
import org.incendo.cloud.velocity.VelocityCommandManager

object PluginCommand {
    fun register(manager: VelocityCommandManager<VelocityAudience>) {
        manager.command(
            manager.commandBuilder("vkick")
                .permission("$ID.vkick")
                .required("target", VelocityPlayersParser.playersParser())
                .optional("reason", StringParser.greedyStringParser())
                .handler { ctx ->
                    val target = ctx.get<VelocityPlayers>("target")
                    val reason = ctx.optional<String>("reason")
                    target.players.forEach {
                        val player = it.source as Player
                        if (!reason.isPresent) {
                            player.disconnect(Component.translatable("multiplayer.disconnect.kicked"))
                        } else {
                            player.disconnect(Component.translatable(reason.get()))
                        }
                    }
                }
        )

        val vgetBuilder = manager.commandBuilder("vget")
            .permission("$ID.vget")
        manager.command(
            vgetBuilder
                .literal("brand")
                .required("target", VelocityPlayersParser.playersParser())
                .handler { ctx ->
                    playerInfoMod(
                        ctx,
                        { it.clientBrand },
                        "no brand"
                    )
                }
        ).command(
            vgetBuilder
                .literal("virtualhost")
                .required("target", VelocityPlayersParser.playersParser())
                .handler { ctx ->
                    playerInfoMod(
                        ctx,
                        { it.virtualHost.getOrNull() },
                        "no virtualhost",
                        { player, info ->
                            if (info.address != null) {
                                "${info.address.hostAddress}:${info.port}"
                            } else {
                                "${info.hostString}:${info.port}"
                            }
                        }
                    )
                }
        ).command(
            vgetBuilder
                .literal("address")
                .required("target", VelocityPlayersParser.playersParser())
                .handler { ctx ->
                    playerInfo(ctx) { "${it.remoteAddress}" }
                }
        ).command(
            vgetBuilder
                .literal("locale")
                .required("target", VelocityPlayersParser.playersParser())
                .handler { ctx ->
                    playerInfo(ctx) { "${it.effectiveLocale}" }
                }
        ).command(
            vgetBuilder
                .literal("version")
                .required("target", VelocityPlayersParser.playersParser())
                .handler { ctx ->
                    playerInfo(ctx) { "${it.protocolVersion}" }
                }
        ).command(
            vgetBuilder
                .literal("vvinfo")
                .required("target", VelocityPlayersParser.playersParser())
                .handler { ctx ->
                    playerInfoMod(
                        ctx,
                        { PluginVVListener.playersDetails[it] },
                        "not use vv",
                        { player, info -> PluginVVListener.playersChannel[player] + info.versionName }
                    )
                }
        )
    }

    private fun playerInfo(ctx: CommandContext<VelocityAudience>, info: (Player) -> String) {
        playerRichInfo(ctx, { true }, info, "")
    }

    private fun <T> playerInfoMod(
        ctx: CommandContext<VelocityAudience>,
        info: (Player) -> T?,
        fallbackInfo: String,
        infoMod: (Player, T) -> String = { player, info -> info as String }
    ) {
        playerRichInfo(ctx, { info.invoke(it) != null }, { infoMod.invoke(it, info.invoke(it)!!) }, fallbackInfo)
    }

    private fun playerRichInfo(
        ctx: CommandContext<VelocityAudience>,
        condition: (Player) -> Boolean,
        info: (Player) -> String,
        fallbackInfo: String
    ) {
        val target = ctx.get<VelocityPlayers>("target")
        target.players.forEach {
            val fullInfo = if (it.source is Player) {
                val player = it.source as Player
                "${player.username}: " + if (condition.invoke(player)) {
                    info.invoke(player)
                } else {
                    fallbackInfo
                }
            } else {
                fallbackInfo
            }
            ctx.sender().sendMessage(
                Component.text(fullInfo)
            )
        }
    }
}
